modules/core/include/opencv2/core/quaternion.hpp

```diff
@@ -0,0 +1,1949 @@
+         * quaternion at first then to do the transformation.
+         *
+         * The operations is similar as toRotMat3x3
+         * except that the points matrix should have the form
+         * \f[\begin{bmatrix}
+         * 0&0&0&...&0\\
+         * x_0& x_1& x_2&...&x_n\\
+         * y_0& y_1& y_2&...&y_n\\
+         * z_0& z_1& z_2&...&z_n
+         * \end{bmatrix}\f]
+         *
+         * @sa toRotMat3x3
+         */
+        cv::Mat toRotMat4x4(bool assumeUnit=false) const;
+
+        /**
+         * @brief transoform the this quaternion to a Vec<T, 4>
+         *
+         * For example
+         * ```
+         * Quatd q(1,2,3,4);
+         * q.toVec();
+         * ```
+         */
+        cv::Vec<_Tp, 4> toVec() const;
+
+        /**
+         * @brief transoform this quaternion to a Rodrigues vector
+         * @param assumeUnit if true, this quaternon should be a unit quaternion and
+         * this function will save some computations.
+         * Rodtigues vector rVec is defined as:
+         * \f[ rVec = [\tan{\frac{\theta}{2}}v_x, \tan{\frac{\theta}{2}}v_y, \tan{\frac{\theta}{2}}v_z]\f]
+         * where \f$\theta\f$ represents rotate angle, and \f$v\f$ represents the normalized axis.
+         * @note Due to the tangent, the quaternion cannot convert to rotation vector when the rotation angle equals \f$pi + 2kpi\f$ radians or \f$180 + 360k\f$ degree, \f$k = ... -2, -1, 1, 2 ...\f$.
+         */
+        cv::Vec<_Tp, 3> toRodrigues(bool assumeUnit=false)  const;
+
+        /**
+        * @brief get the angle of quaternion, it returns the rotate angle.
+        * @param assumeUnit if true, this quaternon should be a unit quaternion and
+        * this function will save some computations.
+        * \f[\psi = 2 *arccos(\frac{w}{||q||})\f]
+        *
+        * For example
+        * ```
+        * Quatd q(1,2,3,4);
+        * q.getAngle();
+        *
+        * q.normalize().getAngle(true);//same as q.getAngle()
+        * ```
+        */
+        _Tp getAngle(bool assumeUnit=false) const;
+
+        /**
+        * @brief get the axis of quaternion, it returns a vector of length 3
+        * @param assumeUnit if true, this quaternon should be a unit quaternion and
+        * this function will save some computations.
+        * the unit axis \f$\boldsymbol{n}\f$ is defined by
+        * \f[\begin{array}
+        *      \boldsymbol{v} &= \boldsymbol{u} ||\boldsymbol{v}||\\
+        *             &= \boldsymbol{u}||q||sin(\theta)
+        *      \end{array}\f]
+        *
+        * For example
+        * ```
+        * Quatd q(1,2,3,4);
+        * q.getAxis();
+        *
+        * q.normalize().getAxis(true);//same as q.getAxis()
+        * ```
+        */
+        cv::Vec<_Tp, 3> getAxis(bool assumeUnit=false) const;
+
+        /**
+         * @brief return the dot between quaternion \f$q\f$ and this quaternion, i.e.,
+         * \f[p \cdot q = this.w \cdot q.w + this.x \cdot x + this.y \cdot y + this.z \cdot z\f]
```

Comment:
The description can be extended by the following:
"dot(a, b) is a good metric of how close the quaternions are. Indeed, consider the quaternion difference `a^-1 * b`, its real part is dot(a, b). At the same time its real part is equal to `cos(beta/2)` where beta is an angle of rotation between a and b. Therefore, the closer dot(a, b) to 1, the smaller rotation between them.

Reply:
OK. Thanks very much again!